### PR TITLE
Add new `getOriginLocation` and `getContextualLocation` methods to the `Problems`

### PR DESCRIPTION
**Pull request by hegyibalint**
This change retains backward compatibility towards older TAPI clients. The behavior will be that older TAPI users will get origin+attributed locations

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/internal/DefaultSingleProblemEvent.java [hunk 1]
```diff
@@ -78,7 +82,19 @@ public Details getDetails() {
...
     public List<Location> getLocations() {
-        return locations;
+        return ImmutableList
+            .<Location>builder()
+            .addAll(originLocation)
+            .addAll(attributedLocations)
+            .build();
+    }
+
+    public List<Location> getOriginLocation() {
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/internal/DefaultSingleProblemEvent.java:92]: ❌ Should be available on `SingleProblemEvent` interface.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/internal/DefaultSingleProblemEvent.java [hunk 2]
```diff
@@ -78,7 +82,19 @@ public Details getDetails() {
...
+        return originLocation;
+    }
+
+    public List<Location> getAttributedLocations() {
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/internal/DefaultSingleProblemEvent.java:96]: ❌ Should be available on `SingleProblemEvent` interface.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/Problem.java [hunk 3]
```diff
@@ -52,9 +52,20 @@ public interface Problem {
...
+     * <p>
+     * E.g. in case of a failed java compilation, the line number and file name where the compilation failed.
+     */
+    List<ProblemLocation> getOriginLocations();
+
+    /**
+     * Returns additional locations where the problem could be attributed to.
+     * <p>
+     * E.g. in case of a failed java compilation,
+     * which plugin caused the compilation to fail (in this example the 'java' plugin),
```
[the remark this answers is not in the document]
Comment (hegyibalint) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/Problem.java:65]: > In any case the documentation should list an example that exists.

What documentation?

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/internal/DefaultSingleProblemEvent.java (hunk 1 above)
Comment (hegyibalint) [reply platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/internal/DefaultSingleProblemEvent.java]: Pulled up

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/Problem.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (hegyibalint) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/Problem.java:65]: I've removed the documentation at least from the TAPI interfaces.
Excuse my snark, but we don't have any documentation so far; it's unclear to me where to put this.

[the remark this answers is not in the document]
Comment (donat) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/Problem.java:65]: I meant this exact javadoc. It's all fine, except I remember that the java compilation problem reports don't contain any information that it's related to the `java` plugin. My request was to use an example that exists.

[the remark this answers is not in the document]
Comment (hegyibalint) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/Problem.java:65]: I came to agree with you, and we should rather document this in our -- currently pretty skim -- user documentation.

**Review by donat — APPROVED** [review]

**Review by donat — APPROVED** [review]

**Comment by donat** [comment]
@bot-gradle merge

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/Problem.java [hunk 4]
```diff
@@ -52,9 +52,18 @@ public interface Problem {
...
+     */
+    List<ProblemLocation> getOriginLocations();
+
+    /**
+     * Returns additional locations, which can help to understand the problem further.
+     * <p>
+     * Might be empty if there is no meaningful contextual information.
      */
-    List<ProblemLocation> getLocations();
+    List<ProblemLocation> getContextualLocations();
```
Comment (wolfs) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/Problem.java:66]: ❔ For my understanding, we currently don't add any contextual location to a problem, right? Do you already have a PR that does this?

Comment (hegyibalint) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/Problem.java:66]: Not yet, we tackled this separation first.

**Pull request merged**